header/p2p/exchange.go

```diff
@@ -288,3 +259,57 @@ func convertStatusCodeToError(code p2p_pb.StatusCode) error {
 		return fmt.Errorf("unknown status code %d", code)
 	}
 }
+
+// sendMessage opens the stream to the given peers and sends ExtendedHeaderRequest to fetch ExtendedHeaders.
+func sendMessage(
+	ctx context.Context,
+	host host.Host,
+	to peer.ID,
+	protocol protocol.ID,
+	req *p2p_pb.ExtendedHeaderRequest,
+) ([]*p2p_pb.ExtendedHeaderResponse, uint64, uint64, error) {
+	startTime := time.Now()
+	stream, err := host.NewStream(ctx, to, protocol)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	// set stream deadline from the context deadline.
+	// if it is empty, then we assume that it will
+	// hang until the server will close the stream by the timeout.
+	if dl, ok := ctx.Deadline(); ok {
+		if err = stream.SetDeadline(dl); err != nil {
+			log.Debugw("error setting deadline: %s", err)
+		}
+	}
+	// send request
+	_, err = serde.Write(stream, req)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		return nil, 0, 0, err
+	}
+	err = stream.CloseWrite()
+	if err != nil {
+		log.Error(err)
+	}
+	headers := make([]*p2p_pb.ExtendedHeaderResponse, 0)
+	totalRequestSize := uint64(0)
+	for i := 0; i < int(req.Amount); i++ {
+		resp := new(p2p_pb.ExtendedHeaderResponse)
+		msgSize, err := serde.Read(stream, resp)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			stream.Reset() //nolint:errcheck
+			return nil, 0, 0, err
+		}
+		totalRequestSize += uint64(msgSize)
+		headers = append(headers, resp)
+	}
+	duration := time.Since(startTime).Milliseconds()
+	if err = stream.Close(); err != nil {
+		log.Errorw("closing stream", "err", err)
+	}
```

Comment:
Same here, I would return this error

Reply:
If it's not closed properly, it likely means that the written data to the stream was not sent. Think of Close as the final Flush on the buffer as well.

Reply:
Discussed internally: we shouldn't return the error in this case, otherwise we fail the entire request (which was probably succeed)

Reply:
Yeah, I mixed up the read and write sides in the comment.